File: internal/app/snippet.go
package app

import (
	"encoding/base64"
	"encoding/json"
	"regexp"
	"strings"
)

// snippet.go — one message, one line, no source.
//
// The activity inbox cuts its previews here, in Go, before they reach a
// browser that could have flattened them. It used to be a whitespace collapse
// and nothing else, so the newest feature's first screen printed
//
//	![image](concord://attach/v2/20ed697d4012474c24a40a5de4f800133…
//
// at people, and two more rows showed literal backticks around `rules.apply`.
//
// The rule is the same one lib/snippet.js applies on the client: a body that
// IS a card is named, block content becomes a token, everything else is prose
// with its formatting flattened. The two files carry the same table of names
// deliberately — the alert-word matcher already lives twice for the same
// reason — because a label that drifts is a row that says something different
// depending on which half of the app wrote it. `TestSnippetMatchesClientRule`
// pins the shared cases.

var (
	snPollRe     = regexp.MustCompile(`\[poll\]\(concord://poll/v1/([A-Za-z0-9_-]+)\)`)
	snAnnounceRe = regexp.MustCompile(`\[announcement\]\(concord://announce/v1/([A-Za-z0-9_-]+)\)`)
	snGameRe     = regexp.MustCompile(`\[game\]\(concord://game/v1/[A-Za-z0-9_-]+\)`)
	snDoodleRe   = regexp.MustCompile(`\[doodle\]\(concord://doodle/v1/[A-Za-z0-9_-]+\)`)
	snSfxRe      = regexp.MustCompile(`\[sound\]\(concord://sfx/v1/[A-Za-z0-9_-]+\)`)
	// Decorations that ride in FRONT of ordinary words: a send effect, a
	// disappearing timer, a sealed timestamp. Never the message.
	snRiderRe = regexp.MustCompile(`\[(?:fx|eph|ts)\]\(concord://(?:fx|eph|ts)/v1/[A-Za-z0-9_-]+\)`)
	// Attachments, image and file, v1 and v2. Mirrors lib/attachments.js and
	// internal/store's own pair.
	snImageRe  = regexp.MustCompile(`!\[image\]\(concord://attach/v[12]/[^)\s]+\)`)
	snFileRe   = regexp.MustCompile(`\[file\]\(concord://file/v1/[^)\s]+\)`)
	snFenceRe  = regexp.MustCompile("(?s)```.*?(?:```|$)")
	snInlineRe = regexp.MustCompile("`([^`]*)`")
	snImgMdRe  = regexp.MustCompile(`!\[[^\]]*\]\([^)]*\)`)
	snLinkMdRe = regexp.MustCompile(`\[([^\]]*)\]\(([^)]*)\)`)
	snBold3Re  = regexp.MustCompile(`\*\*\*([^*]+)\*\*\*`)
	snBold2Re  = regexp.MustCompile(`\*\*([^*]+)\*\*`)
	snItalicRe = regexp.MustCompile(`\*([^*]+)\*`)
	snUnder3Re = regexp.MustCompile(`___([^_]+)___`)
	snUnder2Re = regexp.MustCompile(`__([^_]+)__`)
	snStrikeRe = regexp.MustCompile(`~~([^~]+)~~`)
	snSpoilRe  = regexp.MustCompile(`\|\|([^|]+)\|\|`)
	snDanglRe  = regexp.MustCompile(`(\*\*|~~|\|\|)`)
	snHeadRe   = regexp.MustCompile(`(?m)^#{1,6}\s+`)
	snQuoteRe  = regexp.MustCompile(`(?m)^\s*>\s?`)
	snListRe   = regexp.MustCompile(`(?m)^\s*(?:[-*+]|\d{1,3}\.)\s+`)
)

// The names a quoted card answers to. Short: this is a label in a one-line
// row, not a description. Kept byte-identical to TOKEN_LABELS in
// lib/snippet.js.
const (
	labelGame   = "🎲 game"
	labelDoodle = "🖌 doodle"
	labelSound  = "🔊 sound"
	labelCode   = "📄 code block"
)

func snB64(s string) []byte {
	b, err := base64.RawURLEncoding.DecodeString(strings.TrimRight(s, "="))
	if err != nil {
		return nil
	}
	return b
}

// snippet is the bit of the message the inbox shows. One line, bounded, cut on
// a rune boundary so a truncated emoji does not become a replacement glyph.
func snippet(body string) string {
	const max = 160
	s := flattenBody(body)
	r := []rune(s)
	if len(r) <= max {
		return s
	}
	return strings.TrimSpace(string(r[:max])) + "…"
}

// flattenBody turns a message body into the one line a reader can read.
func flattenBody(body string) string {
	s := snRiderRe.ReplaceAllString(body, " ")

	// A body that IS a card is named rather than quoted. Poll and announcement
	// carry text worth showing; the rest are their own name. Both decodes can
	// fail on a payload something truncated, which is exactly when it matters
	// that the fallback is a label and not the base64 that is left.
	if m := snPollRe.FindStringSubmatch(s); m != nil {
		var p struct {
			Q string `json:"q"`
		}
		if err := json.Unmarshal(snB64(m[1]), &p); err == nil && p.Q != "" {
			return "📊 " + collapse(p.Q)
		}
		return "📊 Poll"
	}
	if m := snAnnounceRe.FindStringSubmatch(s); m != nil {
		var a struct {
			Body string `json:"body"`
		}
		if err := json.Unmarshal(snB64(m[1]), &a); err == nil && a.Body != "" {
			return "📣 " + collapse(flattenMarkdown(a.Body))
		}
		return "📣 Announcement"
	}
	if snGameRe.MatchString(s) {
		return labelGame
	}
	if snDoodleRe.MatchString(s) {
		return labelDoodle
	}
	if snSfxRe.MatchString(s) {
		return labelSound
	}

	// Block content inside prose becomes a token too. A fenced block flattened
	// into one line is eleven lines of somebody's code with the newlines taken
	// out, which is a preview of nothing.
	s = snFenceRe.ReplaceAllString(s, " "+labelCode+" ")

	// Attachments to a placeholder. The caption, if there is one, is the
	// preview; otherwise the placeholder is.
	hasImage := snImageRe.MatchString(s)
	hasFile := snFileRe.MatchString(s)
	if hasImage || hasFile {
		s = snImageRe.ReplaceAllString(s, " ")
		s = snFileRe.ReplaceAllString(s, " ")
		rest := collapse(flattenMarkdown(s))
		glyph, label := "🖼", "image"
		if hasFile && !hasImage {
			glyph, label = "📎", "file"
		}
		if rest != "" {
			return glyph + " " + rest
		}
		return glyph + " " + label
	}
	return collapse(flattenMarkdown(s))
}

// flattenMarkdown reduces formatting to the words inside it. Paired markers
// only, so snake_case and 3 * 4 survive, then a sweep for the multi-character
// markers a truncated excerpt can leave dangling.
func flattenMarkdown(s string) string {
	s = snInlineRe.ReplaceAllString(s, "$1")
	s = snImgMdRe.ReplaceAllString(s, " ")
	s = snLinkMdRe.ReplaceAllString(s, "$1")
	s = snBold3Re.ReplaceAllString(s, "$1")
	s = snBold2Re.ReplaceAllString(s, "$1")
	s = snItalicRe.ReplaceAllString(s, "$1")
	s = snUnder3Re.ReplaceAllString(s, "$1")
	s = snUnder2Re.ReplaceAllString(s, "$1")
	s = snStrikeRe.ReplaceAllString(s, "$1")
	s = snSpoilRe.ReplaceAllString(s, "$1")
	s = snDanglRe.ReplaceAllString(s, "")
	s = snHeadRe.ReplaceAllString(s, "")
	s = snQuoteRe.ReplaceAllString(s, "")
	s = snListRe.ReplaceAllString(s, "")
	return s
}

func collapse(s string) string {
	return strings.Join(strings.Fields(strings.ReplaceAll(s, "\n", " ")), " ")
}

File: frontend/src/lib/snippet.js
// snippet.js — one message, one line, no source.
//
// Three surfaces quote a message somewhere that will never render it: the
// reply strip above a reply, a search result, and the activity inbox. All
// three used to print whatever the body happened to contain, which in a chat
// app that carries polls, games, doodles, sound recipes and attachments as
// bracketed tokens means printing base64 at people:
//
//   ↩ Amina Sadiq  The fold is the whole trick — here it is in eleven lines:
//   ```js export function
//   ![image](concord://attach/v2/7ff3486b8d32b9edd1a3f4cea4231ae0fd19…
//
// The rule is: a body that IS a card is named, block content becomes a token,
// and everything else is prose with its formatting flattened. Never the
// source.
//
// This is the client half. `internal/app/snippet.go` is the same rule in Go,
// for the inbox, whose snippets are cut before they reach the browser — the
// two files carry the same table of token names deliberately, the way the
// alert-word matcher does, because a name that drifts is a row that says
// something different depending on which half wrote it.

import { previewText } from "./attachments.js";
import { stripMarkdown } from "./forum.js";
import { parsePoll, POLL_RE } from "./polls.js";
import { parseAnnounce, ANNOUNCE_RE } from "./announce.js";
import { parseEventToken, EVENT_RE } from "./eventtoken.js";
import { GAME_RE } from "./games.js";
import { DOODLE_RE } from "./doodle.js";
import { SFX_RE } from "./sfxrecipe.js";

// A fence, closed or left open by a truncated excerpt.
const FENCE_RE = /```[\s\S]*?(?:```|$)/g;

// The attachment grammar, loosely. attachments.js parses tokens STRICTLY — a
// 64-character hex id, a 75-character key — which is right for something that
// is about to fetch bytes and wrong for something that is about to print. A
// token cut in half by an excerpt, or written by a build that added a field,
// fails the strict parse and must still never reach a reader as a URI.
const LOOSE_IMG_RE = /!\[image\]\(concord:\/\/attach\/v[12]\/[^)\s]*\)/g;
const LOOSE_FILE_RE = /\[file\]\(concord:\/\/file\/v1\/[^)\s]*\)/g;

// Decorations that ride in FRONT of ordinary words — a send effect, a
// disappearing timer, a sealed timestamp. Written out here rather than
// composed from stripFx/stripEphemeral/stripTimestamp so this module stays
// importable outside a Svelte build (ephemeral.svelte.js opens with a $state),
// and so it reads as the one line its Go counterpart also is.
const RIDER_RE = /\[(?:fx|eph|ts)\]\(concord:\/\/(?:fx|eph|ts)\/v1\/[A-Za-z0-9_-]+\)/g;

// The names a quoted card answers to. Kept short: this is a label inside a
// one-line strip, not a description.
export const TOKEN_LABELS = {
  game: "🎲 game",
  doodle: "🖌 doodle",
  sound: "🔊 sound",
  code: "📄 code block",
};

// plainSnippet: the readable one-line form of a message body.
// `max` of 0 means "do not cut".
export function plainSnippet(content, max = 0) {
  let c = String(content ?? "");
  if (!c) return "";
  // Decorations that ride in FRONT of ordinary words — a send effect, a
  // disappearing timer, a sealed timestamp. They are never the message.
  c = c.replace(RIDER_RE, " ");

  // A body that is a card gets named rather than quoted. Poll and announcement
  // carry a title worth showing; the rest are their own name. Both parses can
  // fail on a payload the excerpt cut in half, which is exactly when it matters
  // that the fallback is a label and not the base64 that is left.
  const poll = parsePoll(c);
  if (poll) return cap(`📊 ${poll.q || "Poll"}`, max);
  if (POLL_RE.test(c)) return "📊 Poll";
  const ann = parseAnnounce(c);
  if (ann) return cap(`📣 ${stripMarkdown(ann.body).replace(/\s+/g, " ").trim() || "Announcement"}`, max);
  if (ANNOUNCE_RE.test(c)) return "📣 Announcement";
  const evt = parseEventToken(c);
  if (evt) return cap(`📅 ${evt.title || "Event"}`, max);
  if (EVENT_RE.test(c)) return "📅 Event";
  if (GAME_RE.test(c)) return cap(TOKEN_LABELS.game, max);
  if (DOODLE_RE.test(c)) return cap(TOKEN_LABELS.doodle, max);
  if (SFX_RE.test(c)) return cap(TOKEN_LABELS.sound, max);
  const invite = inviteSnippet(c);
  if (invite) return cap(invite, max);

  // Block content inside prose becomes a token too. A fenced block flattened
  // into one line is eleven lines of somebody's code with the newlines taken
  // out, which is not a preview of anything.
  c = c.replace(FENCE_RE, ` ${TOKEN_LABELS.code} `);

  // Attachments to a placeholder, then inline formatting to the words inside
  // it, then whitespace to single spaces.
  const shown = previewText(c)
    .replace(LOOSE_IMG_RE, " 🖼 image ")
    .replace(LOOSE_FILE_RE, " 📎 file ");
  const text = stripMarkdown(shown).replace(/\s+/g, " ").trim();
  return cap(text, max);
}

function inviteSnippet(content) {
  const raw = String(content ?? "").trim();
  if (!raw.startsWith("{") || !raw.includes('"op"')) return "";
  try {
    const n = JSON.parse(raw);
    if (!n?.op) return "";
    const name = String(n.guild || n.what || "invite").trim() || "invite";
    if (n.op === "joined") return `joined ${name}`;
    if (n.op === "offered") return `invite to ${name}`;
  } catch {
    /* not an invite card */
  }
  return "";
}

function cap(s, max) {
  if (!max || s.length <= max) return s;
  return s.slice(0, max).trimEnd() + "…";
}
